testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/broker/KcSamlSpDescriptorTest.java

```diff
@@ -0,0 +1,99 @@
+            .setAttribute(SAMLIdentityProviderConfig.ATTRIBUTE_CONSUMING_SERVICE_INDEX, "15")
+            .update())
+        {
+
+            String spDescriptorString = identityProviderResource.export(null).readEntity(String.class);
+            SAMLParser parser = SAMLParser.getInstance();
+            EntityDescriptorType o = (EntityDescriptorType) parser.parse(new StringInputStream(spDescriptorString));
+            SPSSODescriptorType spDescriptor = o.getChoiceType().get(0).getDescriptors().get(0).getSpDescriptor();
+
+            assertThat(spDescriptor.getAttributeConsumingService().isEmpty(), is(false));
+            assertThat(spDescriptor.getAttributeConsumingService().get(0).getIndex(), is(15));
+        }
+    }
+
+    @Test
+    public void testAttributeConsumingServiceNameInSpMetadata() throws IOException, ParsingException, URISyntaxException {
+        try (Closeable idpUpdater = new IdentityProviderAttributeUpdater(identityProviderResource)
+            .setAttribute(SAMLIdentityProviderConfig.ATTRIBUTE_CONSUMING_SERVICE_NAME, "My Attribute Set")
+            .update())
+        {
+
+            String spDescriptorString = identityProviderResource.export(null).readEntity(String.class);
+            SAMLParser parser = SAMLParser.getInstance();
+            EntityDescriptorType o = (EntityDescriptorType) parser.parse(new StringInputStream(spDescriptorString));
+            SPSSODescriptorType spDescriptor = o.getChoiceType().get(0).getDescriptors().get(0).getSpDescriptor();
+
+            assertThat(spDescriptor.getAttributeConsumingService().isEmpty(), is(false));
+            assertThat(spDescriptor.getAttributeConsumingService().get(0).getServiceName().get(0).getValue(), is("My Attribute Set"));
+        }
+    }
+
+    @Test
+    public void testAttributeConsumingServiceMappersInSpMetadata() throws IOException, ParsingException, URISyntaxException {
+        try (Closeable idpUpdater = new IdentityProviderAttributeUpdater(identityProviderResource)
+            .setAttribute(SAMLIdentityProviderConfig.ATTRIBUTE_CONSUMING_SERVICE_INDEX, "12")
+            .update())
+        {
+            IdentityProviderMapperRepresentation attrMapperEmail = new IdentityProviderMapperRepresentation();
+            attrMapperEmail.setName("attribute-mapper-email");
+            attrMapperEmail.setIdentityProviderMapper(UserAttributeMapper.PROVIDER_ID);
+            attrMapperEmail.setConfig(ImmutableMap.<String,String>builder()
+              .put(IdentityProviderMapperModel.SYNC_MODE, IdentityProviderMapperSyncMode.INHERIT.toString())
+              .put(UserAttributeMapper.ATTRIBUTE_NAME, "email_attr_name")
+              .put(UserAttributeMapper.ATTRIBUTE_FRIENDLY_NAME, "email_attr_friendlyname")
+              .put(UserAttributeMapper.USER_ATTRIBUTE, "email")
+              .build());
+            attrMapperEmail.setIdentityProviderAlias(bc.getIDPAlias());
+
+            identityProviderResource.addMapper(attrMapperEmail);
```

Comment:
This change should be reverted after the test, so that we won't interfere other tests. You can add something like this:

```
String mapperId = identityProviderResource.getMappers()
                    .stream()
                    .filter(mapper -> "attribute-mapper-email".equals(mapper.getName()))
                    .findFirst()
                    .map(IdentityProviderMapperRepresentation::getId)
                    .orElse(null);
getCleanup().addCleanup((Runnable) () -> identityProviderResource.delete(mapperId));
```

Reply:
Uhm, I tried adding the code you suggested, but I get a 404 Not Found while running the cleanup routine:

```
12:59:04,570 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] [KcSamlSpDescriptorTest] testAttributeConsumingServiceMappersInSpMetadata() STARTED
12:59:04,783 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] loading test realms
12:59:04,785 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] importing test realms
12:59:04,786 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] --importing realm: consumer
12:59:05,423 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] --importing realm: provider
12:59:05,692 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] creating user for realm provider
12:59:05,909 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding identity provider to realm consumer
12:59:05,999 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client https://localhost:8543/auth/realms/consumer to realm provider
12:59:06,022 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client http://localhost:8280/sales-post/ to realm consumer
12:59:06,062 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client http://localhost:8280/sales-post/.dot/ted to realm consumer
12:59:06,086 DEBUG [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] adding client broker-app to realm consumer
12:59:06,851 INFO  [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] calling all TestCleanup
12:59:06,863 ERROR [org.keycloak.testsuite.broker.KcSamlSpDescriptorTest] failed cleanup!
javax.ws.rs.NotFoundException: HTTP 404 Not Found
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:247)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.DefaultEntityExtractorFactory$3.extractEntity(DefaultEntityExtractorFactory.java:50)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:154)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
	at com.sun.proxy.$Proxy149.delete(Unknown Source)
	at org.keycloak.testsuite.broker.KcSamlSpDescriptorTest.lambda$testAttributeConsumingServiceMappersInSpMetadata$1(KcSamlSpDescriptorTest.java:91)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at java.base/java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
	at java.base/java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658)
	at org.keycloak.testsuite.util.TestCleanup.executeCleanup(TestCleanup.java:130)
	at org.keycloak.testsuite.AbstractKeycloakTest.afterAbstractKeycloakTest(AbstractKeycloakTest.java:218)
...
```

I also tried to move the call to retrieve the mapperId inside the cleanup routine, but I get the same 404 when retrieving the mapper ID:

```
javax.ws.rs.NotFoundException: HTTP 404 Not Found
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.handleErrorStatus(ClientInvocation.java:247)
	at org.jboss.resteasy.client.jaxrs.internal.ClientInvocation.extractResult(ClientInvocation.java:217)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.extractors.BodyEntityExtractor.extractEntity(BodyEntityExtractor.java:64)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invokeSync(ClientInvoker.java:154)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientInvoker.invoke(ClientInvoker.java:115)
	at org.jboss.resteasy.client.jaxrs.internal.proxy.ClientProxy.invoke(ClientProxy.java:76)
	at com.sun.proxy.$Proxy149.getMappers(Unknown Source)
	at org.keycloak.testsuite.broker.KcSamlSpDescriptorTest.lambda$testAttributeConsumingServiceMappersInSpMetadata$1(KcSamlSpDescriptorTest.java:86)
```

Any idea why?

Reply:
Sorry for the confusion, I should have checked this. The cleanup is not necessary because the realm is removed after each test, that is why you are seeing 404.